pandas/core/ops.py

```diff
@@ -1156,6 +1157,82 @@ def add_flex_arithmetic_methods(cls):
     add_methods(cls, new_methods=new_methods)
 
 
+# ----------------------------------------------------------------------
+# Unary
+def _unary_method(cls, op, special=True):
+    op_name = _get_op_name(op, special)
+
+    def f(self):
+        # TODO: Panel
+        if self.ndim > 2:
+            with np.errstate(all='ignore'):
+                # apply op at once based on current impl
+                new_data = op(self.values)
+            result = self._constructor(new_data, self.items,
+                                       self.major_axis, self.minor_axis)
+            return result.__finalize__(self)
+        elif self.ndim == 2:
+            new_data = {i: op(self.iloc[:, i])
+                        for i in range(len(self.columns))}
```

Comment:
So now this happens column by column, while before it was done on the full 2D array of the DataFrame?

Reply:
Yes. If it should consider internal structure, may better to move logics to `internals` (as in first impl)?

Reply:
I am not sure it is needed to do it block by block, but I am mainly wondering if this would have implications. 

Eg something like 

```
In [1]: df = pd.DataFrame({'a': [1, 2, 3], 'b':[False, True, False]})

In [2]: df
Out[2]: 
   a      b
0  1  False
1  2   True
2  3  False

In [3]: -df
Out[3]: 
    a   b
0  -1   0
1  -2  -1
2  -3   0

In [4]: -df['b']
Out[4]: 
0     True
1    False
2     True
Name: b, dtype: bool
```

has a different behaviour on the full dataframe vs the column (which is not necessarily bad in this case, but we should check)

Reply:
@jorisvandenbossche Yes, the PR fixes unintended dtype changes. `pandas/tests/frame/test_arithmetic.py` has some mixed dtype tests.

```
# Current master
(-pd.DataFrame({'a': [1], 'b': [1.1]})).dtypes
# a    float64
# b    float64
# dtype: object

# After the PR
(-pd.DataFrame({'a': [1], 'b': [1.1]})).dtypes
# a      int64
# b    float64
# dtype: object
```